### PR TITLE
Refined logging

### PR DESCRIPTION
**Pull request by SmilyOrg**

### CODE REVIEW

File: loom/common/assets/assets.cpp [hunk 1]
```diff
@@ -655,7 +677,16 @@ class AssetProtocolFileMessageListener : public AssetProtocolMessageListener
                int lastByteOffset = chunkOffset + fileBitsLength;
 
                // Log it.
-               lmLogInfo(gAssetLogGroup, "FILE '%s' %d of %d bytes!", pendingFilePath.c_str(), lastByteOffset, pendingFileLength);
+               int elapsed = loom_readTimer(pendingFileTimer);
+               if ((pendingFileInit && elapsed > PROGRESS_INIT_TIME) ||
```
Comment (bengarney) [inline loom/common/assets/assets.cpp]: Ah that's nice.

File: loom/common/assets/assets.cpp [hunk 2]
```diff
@@ -724,7 +755,7 @@ static void loom_asset_serviceServer()
         if (loom_net_isSocketDead(gAssetServerSocket) == 1)
         {
             // Might be DOA, ie, connect failed.
-            lmLog(gAssetLogGroup, "Failed to connect to asset server %s:%d", ASSET_STREAM_HOST, ASSET_STREAM_PORT);
+            lmLogWarn(gAssetLogGroup, "Failed to connect to asset server %s:%d", ASSET_STREAM_HOST, ASSET_STREAM_PORT);
```
Comment (bengarney) [inline loom/common/assets/assets.cpp]: Should it be an error?

File: loom/common/assets/assets.cpp [hunk 3]
```diff
@@ -842,7 +873,8 @@ void loom_asset_pump()
       else
       {
         // Instate the asset.
-        asset->instate(type, assetBits, dtor);        
+        asset->instate(type, assetBits, dtor);
+        asset->blob->length = size;
```
Comment (bengarney) [inline loom/common/assets/assets.cpp]: Why is this needed?

File: loom/common/assets/assets.cpp [hunk 4]
```diff
@@ -1016,15 +1048,17 @@ void *loom_asset_lock(const char *name, unsigned int type, int block)
...
     }
 
+    bool preloaded = asset->state == loom_asset_t::Loaded;
+
     // Otherwise, let's force it to load now.
-    if (asset->state != loom_asset_t::Loaded)
+    if (!preloaded)
     {
-        lmLogDebug(gAssetLogGroup, "Blocking so forcing load of '%s'.", namePtr);
+        lmLogDebug(gAssetLogGroup, "Loading '%s'", namePtr);
```
Comment (bengarney) [inline loom/common/assets/assets.cpp]: Should we preserve a little more info about the event in the log msg?

File: loom/common/assets/assets.cpp [hunk 5]
```diff
@@ -1032,13 +1066,13 @@ void *loom_asset_lock(const char *name, unsigned int type, int block)
...
         {
-            lmLogDebug(gAssetLogGroup, "Pumping load of '%s'...", namePtr);
+            lmLogDebug(gAssetLogGroup, "Pumping load of '%s'", namePtr);
             loom_asset_pump();
         }
 
         if (asset->state != loom_asset_t::Loaded)
         {
-            lmLogError(gAssetLogGroup, "Unable to load asset '%s'!", name);
+            lmLogError(gAssetLogGroup, "Load failed for '%s'!", name);
```
Comment (bengarney) [inline loom/common/assets/assets.cpp]: Failed to load asset '%s'?

File: loom/common/assets/assets.cpp [hunk 6]
```diff
@@ -1057,7 +1091,14 @@ void *loom_asset_lock(const char *name, unsigned int type, int block)
 
     loom_mutex_unlock(gAssetLock);
 
-    lmLogInfo(gAssetLogGroup, "Locked asset '%s'...", namePtr);
+    if (preloaded)
+    {
+        lmLogDebug(gAssetLogGroup, "Acquired '%s'", namePtr);
```
Comment (bengarney) [inline loom/common/assets/assets.cpp]: Ah, nice.

File: loom/common/assets/assetsSound.cpp [hunk 7]
```diff
@@ -31,7 +31,7 @@
 #include "wavloader.h"
 
 extern "C" loom_allocator_t *gAssetAllocator;
-loom_logGroup_t gSoundAssetGroup = { "soundAsset", 1 };
+loom_logGroup_t gSoundAssetGroup = { "asset.sound", 1 };
```
Comment (bengarney) [inline loom/common/assets/assetsSound.cpp]: Good catch.

File: loom/common/assets/assetsSound.cpp [hunk 8]
```diff
@@ -216,6 +216,6 @@ void *loom_asset_soundDeserializer( void *buffer, size_t bufferLen, LoomAssetCle
       return 0;
    }
 
-    lmLogInfo(gSoundAssetGroup, "Allocated %d bytes for a sound!", sound->bufferSize);
+    lmLogDebug(gSoundAssetGroup, "Allocated %d bytes for a sound!", sound->bufferSize);
```
Comment (bengarney) [inline loom/common/assets/assetsSound.cpp]: Should match image allocation above?

File: loom/common/core/log.c [hunk 9]
```diff
@@ -225,12 +232,28 @@ void loom_log_addRule(const char *prefix, int enabled, int filterLevel)
...
+    return
+        strcmp(level, "debug") == 0 ? LoomLogDebug :
+        strcmp(level, "verbose") == 0 ? LoomLogDebug :
+        strcmp(level, "info") == 0 ? LoomLogInfo :
+        strcmp(level, "warn") == 0 ? LoomLogWarn :
+        strcmp(level, "warning") == 0 ? LoomLogWarn :
+        strcmp(level, "error") == 0 ? LoomLogError :
+        strcmp(level, "quiet") == 0 ? LoomLogNone :
+        strcmp(level, "none") == 0 ? LoomLogNone :
+        strcmp(level, "default") == 0 ? LoomLogDefault :
```
Comment (bengarney) [inline loom/common/core/log.c]: Probably should update markdown docs to capture this?

File: loom/common/utils/humanFormat.h [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+ * ===========================================================================
+ */
+
+#ifndef _HUMAN_FORMAT_H_
+#define _HUMAN_FORMAT_H_
+
+#include "math.h"
+#include "loom/common/utils/utString.h"
+
+utString humanFileSize(long bytes)
```
Comment (bengarney) [inline loom/common/utils/humanFormat.h]: Oh nice, is this based on anything?

File: loom/engine/bindings/loom/lmApplication.cpp [hunk 11]
```diff
@@ -136,26 +146,51 @@ int LoomApplication::initializeTypes()
...
 
-void LoomApplication::execMainAssembly()
+void LoomApplication::initMainAssembly()
 {
     lmAssert(!rootVM, "VM already running");
     rootVM = lmNew(NULL) LSLuaState();
+
+    ensureInitialAssetSystem();
+
+    //*
```
Comment (bengarney) [inline loom/engine/bindings/loom/lmApplication.cpp]: Dead comments?

File: loom/engine/bindings/loom/lmApplicationTasks.cpp [hunk 12]
```diff
@@ -51,7 +51,7 @@ void loom_tick()
 
     performance_tick();
 
-    profilerBlock_t p = { "loom_tick", platform_getMilliseconds(), 8 };
+    profilerBlock_t p = { "loom_tick", platform_getMilliseconds(), 17 };
```
Comment (bengarney) [inline loom/engine/bindings/loom/lmApplicationTasks.cpp]: Why is 17 the right threshold here?

File: loom/engine/bindings/loom/lmHTTPRequest.cpp [hunk 13]
```diff
@@ -75,7 +75,7 @@ class HTTPRequest {
         // Warn if this object is garbage collected while a requet is pending
         if (requestPending)
         {
-            lmLog(gHTTPRequestLogGroup, "WARNING: HTTPRequest object with url \"%s\" garbage collected before completing a pending request. Ensure references to HTTPRequest objects are maintained while requests are pending.", url.c_str())
+            lmLogWarn(gHTTPRequestLogGroup, "HTTPRequest object with url \"%s\" garbage collected before completing a pending request. Ensure references to HTTPRequest objects are maintained while requests are pending.", url.c_str())
```
Comment (bengarney) [inline loom/engine/bindings/loom/lmHTTPRequest.cpp]: This is almost always a bug, is it going to be visible by default?

File: loom/engine/sound/lsSound.cpp [hunk 14]
```diff
@@ -87,7 +87,7 @@ extern "C"
         alDistanceModel(AL_INVERSE_DISTANCE_CLAMPED);
         CHECK_OPENAL_ERROR();
 
-        lmLogInfo(gLoomSoundLogGroup, "Loom Sound engine OpenAL '%s' initialized.", alcGetString(dev, ALC_ALL_DEVICES_SPECIFIER));
+        lmLogDebug(gLoomSoundLogGroup, "Initialized sound '%s'", alcGetString(dev, ALC_ALL_DEVICES_SPECIFIER));
```
Comment (bengarney) [inline loom/engine/sound/lsSound.cpp]: Should this say "sound device" or something to indicate it's not a specific sound?

File: loom/graphics/gfxGraphics.cpp [hunk 15]
```diff
@@ -324,26 +322,26 @@ void Graphics::handleContextLoss()
...
+    lmLogDebug(gGFXLogGroup, "Handle context loss: Done");
 }
 
 
 void Graphics::screenshot(const char *path)
 {
-    if (strlen(path) > 1024) {
-        lmLog(gGFXLogGroup, "Screenshot name too big! Screenshots must be 1024 characters or less");
+    if (strlen(path) > 1023) {
+        lmLog(gGFXLogGroup, "Screenshot path too long! Screenshot path must be 1023 characters or fewer");
```
Comment (bengarney) [inline loom/graphics/gfxGraphics.cpp]: Is this an error?

File: loom/graphics/gfxVectorRenderer.cpp [hunk 16]
```diff
@@ -596,7 +596,7 @@ void VectorSVG::resetImage() {
 }
 
 void VectorSVG::loadFile(utString path, utString units, float dpi) {
-    lmLogError(gGFXVectorRendererLogGroup, "Loading %s", path.c_str());
+    lmLogDebug(gGFXVectorRendererLogGroup, "Loading '%s'", path.c_str());
```
Comment (bengarney) [inline loom/graphics/gfxVectorRenderer.cpp]: Good.

File: loom/script/native/lsNativeDelegate.cpp [hunk 17]
```diff
@@ -28,7 +28,7 @@
 #include "loom/script/native/lsNativeDelegate.h"
 
 
-lmDefineLogGroup(gNativeDelegateGroup, "script.native", 1, LoomLogInfo);
+lmDefineLogGroup(gNativeDelegateGroup, "script.ntv", 1, LoomLogInfo);
```
Comment (bengarney) [inline loom/script/native/lsNativeDelegate.cpp]: Maybe just do "delegate"? ntv is pretty terse.

File: loom/script/native/lsNativeInterface.cpp [hunk 18]
```diff
@@ -26,6 +26,8 @@
 #include "loom/script/reflection/lsPropertyInfo.h"
 #include "loom/common/core/allocator.h"
 
+lmDefineLogGroup(nativeInterfaceLogGroup, "script.itf", 1, LoomLogInfo);
```
Comment (bengarney) [inline loom/script/native/lsNativeInterface.cpp]: Maybe interface? itf is also super terse.

File: loom/script/runtime/lsLuaState.cpp [hunk 19]
```diff
@@ -591,21 +620,38 @@ Assembly *LSLuaState::loadExecutableAssemblyBinary(const char *buffer, long buff
     lmCheck(headerBytes.readUnsignedInt() == LOOM_BINARY_VERSION_MINOR, "minor version mismatch");
     unsigned int sz = headerBytes.readUnsignedInt();
 
-    utByteArray bytes;
-    bytes.resize(sz);
+    utByteArray *bytes = lmNew(NULL) utByteArray();
```
Comment (bengarney) [inline loom/script/runtime/lsLuaState.cpp]: Does this avoid a copy op?

File: tools/assetAgent/src/main.cpp [hunk 20]
```diff
@@ -558,7 +558,7 @@ static void processFileEntryDeltas(utArray<FileEntryDelta> *deltas)
         // Note we are using gActiveHandlers.size() outside of a lock, but this is ok as it's a word.
         if ((strstr(canonicalFile, ".loom") || strstr(canonicalFile, ".ls")) && (gActiveHandlers.size() > 0))
         {
-            lmLog(gAssetAgentLogGroup, "SENDING       -  %s", canonicalFile);
+            lmLog(gAssetAgentLogGroup, "Changed '%s'", canonicalFile);
```
Comment (bengarney) [inline tools/assetAgent/src/main.cpp]: Why shouldn't we indicate what we are sending?

File: tools/assetAgent/src/telemetryServer.cpp [hunk 21]
```diff
@@ -191,7 +191,7 @@ void TelemetryServer::start()
     // Set /stream as the websocket connection that streams all the ticks
     mg_set_websocket_handler(server, LTS_STREAM_URI, StreamConnectHandler, StreamReadyHandler, StreamDataHandler, StreamCloseHandler, NULL);
 
-    lmLog(gTelemetryServerLogGroup, "Loom Telemetry server listening on port %s", mg_get_option(server, "listening_ports"));
+    lmLog(gTelemetryServerLogGroup, "Loom Telemetry server running at http://localhost:%s/", mg_get_option(server, "listening_ports"));
```
Comment (bengarney) [inline tools/assetAgent/src/telemetryServer.cpp]: Good clarification.

File: tools/ldb/src/main.cpp [hunk 22]
```diff
@@ -53,7 +53,7 @@ int main(int argc, const char** argv) {
     debuggerVM = new LSLuaState();
     debuggerVM->open();
 
-    lmLog(ldbLogGroup,"o executing LDB");
+    lmLog(ldbLogGroup, "Running");
```
Comment (bengarney) [inline tools/ldb/src/main.cpp]: Why is this better?

File: tools/lsc/src/main.cpp [hunk 23]
```diff
@@ -178,18 +204,41 @@ int main(int argc, const char **argv)
...
     const char *sdkRoot = NULL;
+
+    for (int i = 1; i < argc; i++)
+    {
+        if (!strcmp(argv[i], "--log-type"))
+        {
+            i++;
+            if (i >= argc)
+            {
+                LSError("--log-type option requires the log type (default|cli|runtime) to be specified next");
```
Comment (bengarney) [inline tools/lsc/src/main.cpp]: What does this option do?

**Comment by bengarney** [comment]
Nitted!

File: loom/common/assets/assets.cpp (hunk 2 above)
Comment (SmilyOrg) [inline loom/common/assets/assets.cpp]: Hmm, ok.

File: loom/common/assets/assets.cpp (hunk 3 above)
Comment (SmilyOrg) [inline loom/common/assets/assets.cpp]: So we can display the loaded size later on, that field wasn't used / set before.

File: loom/common/assets/assets.cpp (hunk 4 above)
Comment (SmilyOrg) [inline loom/common/assets/assets.cpp]: That's really the only place loading happens. If blocking isn't enabled and loading fails, it has a more descriptive error message. I think adding implementation details in the log message where it isn't necessary for differentiation only serves to confuse. If you really want to know what's going on it's pretty easy to just search for `"Loading ` and look at the implementation.

File: loom/common/assets/assets.cpp (hunk 5 above)
Comment (SmilyOrg) [inline loom/common/assets/assets.cpp]: :+1:

File: loom/common/assets/assetsSound.cpp (hunk 8 above)
Comment (SmilyOrg) [inline loom/common/assets/assetsSound.cpp]: Ah, missed it, thanks!

File: loom/common/core/log.c (hunk 9 above)
Comment (SmilyOrg) [inline loom/common/core/log.c]: Yes.

File: loom/common/utils/humanFormat.h (hunk 10 above)
Comment (SmilyOrg) [inline loom/common/utils/humanFormat.h]: It's based on me not finding any obvious small lib for this and making one. Wanted to just copy something like http://programanddesign.com/cpp/human-readable-file-size-in-c/

But mine is nicer :smirk_cat:

File: loom/engine/bindings/loom/lmApplication.cpp (hunk 11 above)
Comment (SmilyOrg) [inline loom/engine/bindings/loom/lmApplication.cpp]: Yep.

File: loom/engine/bindings/loom/lmApplicationTasks.cpp (hunk 12 above)
Comment (SmilyOrg) [inline loom/engine/bindings/loom/lmApplicationTasks.cpp]: 1000/60 = 16.666..., so 17 is that with some buffer. It's not variable like it probably should be, but it's better than 8 at least, since with 8 it just printed that it was over threshold every frame (since with 60fps vsync every frame is obviously going to be >8ms).

File: loom/engine/bindings/loom/lmHTTPRequest.cpp (hunk 13 above)
Comment (SmilyOrg) [inline loom/engine/bindings/loom/lmHTTPRequest.cpp]: Yes. Everything >= Info is visible by default. Of course you can explicitly filter to only see errors, in which case this won't show, but that's your fault.

File: loom/engine/sound/lsSound.cpp (hunk 14 above)
Comment (SmilyOrg) [inline loom/engine/sound/lsSound.cpp]: Good point.

File: loom/graphics/gfxGraphics.cpp (hunk 15 above)
Comment (SmilyOrg) [inline loom/graphics/gfxGraphics.cpp]: I guess yeah.

File: loom/script/native/lsNativeDelegate.cpp (hunk 17 above)
Comment (SmilyOrg) [inline loom/script/native/lsNativeDelegate.cpp]: Ok.

Comment (SmilyOrg) [inline loom/script/native/lsNativeDelegate.cpp]: Or.. maybe we can just do `ls` instead of `script` and then we can have `ls.native` and `ls.iface`?

File: loom/script/runtime/lsLuaState.cpp (hunk 19 above)
Comment (SmilyOrg) [inline loom/script/runtime/lsLuaState.cpp]: I think we didn't have to return the bytearray before, so we just used a local, but now loading the bytes and loading the assembly are separate operations, so we need a way to get the bytearray and just returning a local would copy it afaik?

File: tools/assetAgent/src/main.cpp (hunk 20 above)
Comment (SmilyOrg) [inline tools/assetAgent/src/main.cpp]: I think only displaying it as changed makes it more meaningful to the user. The fact that it's sending it over an asset agent connection seems to me more like an implementation detail than anything.

If you change a file, it makes more sense to see "Changed" pop up in the log rather than "Sending". Now sending does convey the additional meaning that the operation will take a bit to take hold, but that should be covered by the "Updating/Updated" messages that appear right after anyway.

After a couple more tweaks, the reload log looks like this now (first updating `credits.png` then `Thrusties.ls`):

```
     asset  Updated 'assets\credits.png', 78.3 KiB (80145 bytes)
     agent  Changed 'src\Thrusties.ls'
       cli  Recompiling 'src\Thrusties.ls'
     asset  Updated 'src\Thrusties.ls', 2.4 KiB (2445 bytes)
  compiler  LSC - Release JIT Compiler
  compiler  SDK: C:\Users\Miha\.loom\sdks\dev\
  compiler  Compiling: Main
  compiler  Linking: Feathers
  compiler  Linking: Loom
  compiler  Linking: System
  compiler  Compile successful: .\bin\Main.loom
     agent  Changed 'bin\Main.loom'
     asset  Updating 'bin\Main.loom', 0.8 MiB (858743 bytes)
     asset    12%
     asset    41%
     asset    71%
     asset  Updated 'bin\Main.loom', 0.8 MiB (858743 bytes)
       app  Reloading Main.loom
       app  ---------
       app  Thrusties
       app  ---------
     asset  Loaded 'assets/splash.png', 5.0 KiB (5141 bytes)
     asset  Loaded 'assets/bg.png', 2.7 KiB (2760 bytes)
```

What do you think?

File: tools/assetAgent/src/telemetryServer.cpp (hunk 21 above)
Comment (SmilyOrg) [inline tools/assetAgent/src/telemetryServer.cpp]: :+1:

File: tools/ldb/src/main.cpp (hunk 22 above)
Comment (SmilyOrg) [inline tools/ldb/src/main.cpp]: Because

```
       ldb  Running
```

seems nicer / less redundant than

```
       ldb  o executing LDB
```

It fits the surrounding messages more as well.

File: tools/lsc/src/main.cpp (hunk 23 above)
Comment (SmilyOrg) [inline tools/lsc/src/main.cpp]: It toggles between output as a cli tool (just normal output) and output as a runtime asset agent logger (prefixed with `  compiler  ` so it fits along with other messages).

File: loom/script/native/lsNativeInterface.cpp (hunk 18 above)
Comment (SmilyOrg) [inline loom/script/native/lsNativeInterface.cpp]: It is kinda ugly. Hopefully you never see the messages from this and above anyway though :)

**Comment by SmilyOrg** [comment]
What's up with this? :P
